### PR TITLE
EventDB: allocation table integration test fix

### LINKED ISSUE #763: Event db: Allocation table
We need a new Event database Allocation table to mirror [storagesc.StorageAllocaion](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/models.go#L544)

We need code in event database to include the table as well as support events that add, delete or update allocations. Also we need to take account of the extra complications of sub tables.

TransactionId, foreign key into Transcation table, is the last transaction to modify the allocation object.
[Terms](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/models.go#L404) , the variable data from the StorateAllocaion.BlobberDetails.Terms object.
Curators, a []string field in the StorageAllocastion object
ReadPools, allocation pools with IsWritePool = false
WrietPools, allocation pools with IsWritePool = true

We need to emit events in `storagesc` when an allocation is added or updated. changes:
[newAllocationRequest](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/allocation.go#L401)
[UpdateAlloctionRequest](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/allocation.go#L1057)
[Finilise allocation](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/allocation.go#L1341)
[Cancel allocation](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/allocation.go#L1267)

The [\allocation](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/handler.go#L178) can be redone to use this table. The [\allocations](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/handler.go#L106) also, but this should be appreciated or redesigned.

| Event db Allocaion table   | type             | StorageAllocation              | type               | notes                                                     |
|----------------------------|------------------|--------------------------------|--------------------|-----------------------------------------------------------|
| AllocaionID                | string           | ID                             | string             | unique key                                                |
| TransactionId              | uint             | Tx                             | string             | foreign key into transaction table                        |
| DataShards                 | int              | DataShards                     | int                |                                                           |
| ParityShards               | int              | ParityShards                   | int                |                                                           |
| Size                       | int64            | Size                           | int64              |                                                           |
| Expiration                 | int64            | Expiration                     | Timestamp          |                                                           |
| Terms                      | []Terms          | BlobberDetails                 | []BlobberAllocaion | Terms sub table contains foreign keys for linked blobbers |
| Owner                      | string           | Owner                          | string             | Should this be foreign key into user table?               |
| OwnerPublicKey             | string           | OwnerPublicKey                 | string             |                                                           |
| IsImmutable                | bool             | IsImmutable                    | bool               |                                                           |
| ReadPriceMin               | int64            | ReadPriceRange.Min             | Balance            |                                                           |
| ReadPriceMax               | inat64           | ReadPriceRange.Max             | Balance            |                                                           |
| WritePriceMin              | int64            | WritePriceRange.Min            | Balance            |                                                           |
| WritePriceMax              | int64            | WritePriceRange.Max            | Balance            |                                                           |
| MaxChallengeCompletionTime | int64            | MaxChallengeCompletionTime     | time.Duration      |                                                           |
| ChallengeCompletionTime    | int64            | ChallengeCompletionTime        | time.Duration      |                                                           |
| StartTime                  | int64            | StartTime                      | TimeStamp          |                                                           |
| Finalized                  | bool             | Finalized                      | bool               | Only one of finalized or cancelled can be true.           |
| Cancelled                  | bool             | Canceled                       | bool               | Misspelt                                                  |
| UsedSize                   | int64            | UsedSize                       | int64              |                                                           |
| MovedToChallenge           | int64            | MovedToChallenge               | Balance            |                                                           |
| MovedBack                  | int64            | MovedBack                      | Balance            |                                                           |
| MovedToValidators          | int64            | MovedToValidators              | Balance            |                                                           |
| Curators                   | []Curator        | Curators                       | []string           | curator sub table                                         |
| TimeUnit                   | int64            | TimeUnit                       | time.Duration      |                                                           |
| UsedSize                   | int64            | Stats.UsedSize                 | int64              | Stats will be moved out of StorageAllocation object       |
| NumWrites                  | int64            | Stats.NumWrites                | int64              | Stats will be moved out of StorageAllocation object       |
| NumReads                   | int64            | Stats.NumReads                 | int64              | Stats will be moved out of StorageAllocation object       |
| TotalChallenges            | int64            | Stats.TotalChallenges          | int64              | Stats will be moved out of StorageAllocation object       |
| OpenChallenges             | int64            | Stats.OpenChallenges           | int64              | Stats will be moved out of StorageAllocation object       |
| SuccesfulChallenges        | int64            | Stats.SuccessChallenges        | int64              | Stats will be moved out of StorageAllocation object       |
| FailedChallenges           | int64            | Stats.FailedChallenges         | int64              | Stats will be moved out of StorageAllocation object       |
| LatestClosedChallengeTxn   | uint             | Stats.LatestClosedChallengeTxn | string             | foreign key into Transaction table                        |

### PR DESCRIPTION
**Pull request by iamrz1**
## Fixes

Allocation Table

### CODE REVIEW

File: code/go/0chain.net/smartcontract/dbs/event/allocation.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	var count int64
+	result := edb.Store.Get().Model(&Allocation{}).Where("finalized = ? AND cancelled = ?", false, false).Count(&count)
+	if result.Error != nil {
+		return 0, fmt.Errorf("error retrieving active allocations , error: %v", result.Error)
+	}
+
+	return count, nil
+}
+
+func (edb EventDb) GetBlobberAllocationCount() (int64, error) {
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/dbs/event/allocation.go]: is this active allocations blobber count?
if yes, name it GetActiveAllocsBlobberCount()

File: code/go/0chain.net/smartcontract/storagesc/challenge.go [hunk 2]
```diff
@@ -762,6 +774,12 @@ func (sc *StorageSmartContract) addChallenge(alloc *StorageAllocation,
 	if err != nil {
 		return "", err
 	}
+
+	err = emitAddOrOverwriteAllocation(alloc, balances)
+	if err != nil {
+		return "", common.NewErrorf("blobber_challenge_decode_error",
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/storagesc/challenge.go]: blobber_challenge_decode_error ?

**Review by ssardana96 — CHANGES_REQUESTED** [review]
lgtm. just some small neats

**Comment by ssardana96** [comment]
@Sriep can you also review once.

File: code/go/0chain.net/smartcontract/storagesc/challenge.go (hunk 2 above)
Comment (iamrz1) [reply code/go/0chain.net/smartcontract/storagesc/challenge.go]: yes, this seems unusual, I'll return a relevant error.

File: code/go/0chain.net/smartcontract/dbs/event/allocation.go (hunk 1 above)
Comment (iamrz1) [reply code/go/0chain.net/smartcontract/dbs/event/allocation.go]: it look like the method `GetBlobberAllocationCount` is aptly named after the name of the handler `blobber_allocations_count` but the sql command suggests it should be `GetActiveAllocsBlobberCount`.  I'll change it to `GetActiveAllocsBlobberCount` for now.

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 3]
```diff
@@ -1400,6 +1420,12 @@ func (sc *StorageSmartContract) finalizeAllocation(
 			"saving allocation: "+err.Error())
 	}
 
+	err = emitAddOrOverwriteAllocation(alloc, balances)
+	if err != nil {
+		return "", common.NewErrorf("alloc_Cancel_failed",
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go]: `alloc_Cancel_failed` -> `alloc_cancel_failed`

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/storagesc/allocation.go (hunk 3 above)
Comment (iamrz1) [reply code/go/0chain.net/smartcontract/storagesc/allocation.go]: fixed and rebased.

**Review by peterlimg — DISMISSED** [review]
lgtm

**Review by Sriep — DISMISSED** [review]

**Review by ssardana96 — APPROVED** [review]

**Pull request merged**